src/lib.rs

```diff
@@ -0,0 +1,71 @@
+#![cfg_attr(not(test), no_std)]
+
+//! Gimlet Inspector protocol definition.
+//!
+//! The types in this file are intended to be encoded with `hubpack` and stuffed
+//! into a UDP packet. Messages to Gimlet use the `Request` type. Responses from
+//! Gimlet use the `*Response` types -- a response structure may be specific to
+//! the request. In both cases, certain requests/responses may append binary
+//! data _after_ the `hubpack`-encoded data. This is documented below on the
+//! specific items.
+
+use hubpack::SerializedSize;
+use serde::{Serialize, Deserialize};
+
+/// Request format to the inspector agent.
+///
+/// This is an enum so that we implicitly get a protocol version field at the
+/// start of all encoded packets. In the event that multiple versions coexist at
+/// one time, we can add multiple variants. As versions are obsoleted, we can
+/// replace them with stub variants to ensure their indexes don't get reused.
+///
+/// The downside of this approach is that it will fail at 256 versions.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, SerializedSize)]
+pub enum Request {
+    /// A request in v0 consists only of the name of the query to be issued.
+    V0(Query),
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, SerializedSize)]
+pub enum Query {
+    /// Asks the agent to interrogate the sequencer FPGA and send the register
+    /// contents back. The response is always a `SequencerRegistersResponse`.
+    SequencerRegisters,
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, SerializedSize)]
+pub enum SequencerRegistersResponse {
+    /// The agent successfully contacted the sequencer and collected its
+    /// registers. They are appended in the binary payload section of the
+    /// message. The number of bytes appended may depend on the sequencer
+    /// revision, but the sequencer revision is always in the first bytes of the
+    /// registers. At the time of this writing, 64 bytes will be appended.
+    Success,
+
+    /// The agent was unable to contact the sequencer task because it crashed
+    /// during the attempt. No data is attached.
+    SequencerTaskDead,
+
+    /// The agent contacted the sequencer task, but _it_ was unable to contact
+    /// the FPGA. The only way this can fail in the current sequencer is the
+    /// `ReadRegsFailure` variant, which is implied here. No data is attached.
+    SequencerReadRegsFailed,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn v1_encoding_check() {
+        let message = Request::V0(Query::SequencerRegisters);
+        let mut encoded = [0; Request::MAX_SIZE];
+        let len = hubpack::serialize(&mut encoded, &message).unwrap();
+
```

Comment:
If `Query` is meant to be closed (see [this comment](https://github.com/oxidecomputer/gimlet-inspector-protocol/pull/1/files#r1466442639)), then we should add something that fails to compile if new members are added, e.g.
```rust
// Dummy match statement to prevent people from adding more items
match Query::SequencerRegisters {
    Query::SequencerRegister => (),
    // Don't add more variants to the enum!
}
```
(is there a better way to do this?)

Reply:
Query's not meant to be closed. I added notes on what to do if you add a query.